tool/tsh/kubectl.go

```diff
@@ -275,3 +312,158 @@ func getKubeClusterName(args []string, teleportClusterName string) (string, erro
 	}
 	return kubeName, nil
 }
+
+type kubeLocalProxyOpts struct {
+	kubectlArgs                    []string
+	makeAndStartKubeLocalProxyFunc func(*CLIConf, *clientcmdapi.Config, kubeconfig.LocalProxyClusters) (func(), string, error)
+	setEnvFunc                     func(string, string) error
+}
+
+type applyKubeLocalProxyOpts func(o *kubeLocalProxyOpts)
+
+func withKubectlArgs(args []string) applyKubeLocalProxyOpts {
+	return func(o *kubeLocalProxyOpts) {
+		o.kubectlArgs = args
+	}
+}
+
+func newKubeLocalProxyOpts(applyOpts ...applyKubeLocalProxyOpts) kubeLocalProxyOpts {
+	opts := kubeLocalProxyOpts{
+		makeAndStartKubeLocalProxyFunc: makeAndStartKubeLocalProxy,
+		setEnvFunc:                     os.Setenv,
+	}
+	for _, applyOpt := range applyOpts {
+		applyOpt(&opts)
+	}
+	return opts
+}
+
+// maybeStartKubeLocalProxy starts a kube local proxy if local proxy is
+// required, and overwrites "KUBECONFIG" environment variable to the kubeconfig
+// generated by the local proxy. Called by `tsh kubectl` and `tsh kube exec`.
+func maybeStartKubeLocalProxy(cf *CLIConf, applyOpts ...applyKubeLocalProxyOpts) (func(), error) {
+	opts := newKubeLocalProxyOpts(applyOpts...)
+
+	config, clusters, useLocalProxy := shouldUseKubeLocalProxy(cf, opts.kubectlArgs)
+	if !useLocalProxy {
+		return func() {}, nil
+	}
+
+	closeFn, newKubeconfigLocation, err := opts.makeAndStartKubeLocalProxyFunc(cf, config, clusters)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Update KUBECONFIG path.
+	if err := opts.setEnvFunc("KUBECONFIG", newKubeconfigLocation); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	overwriteKubeconfigFlagInArgs(opts.kubectlArgs, newKubeconfigLocation)
```

Comment:
Not sure if this logic responsible for setting env  and overwriting CLI flag should be part of  the `maybeStartKubeLocalProxy` function.

Reply:
I tried today to move this logic outside but I found it a little messy to return more values. I've ended up changing this to be a generic option, and the caller has to explicitly set it like this:
```
closeFn, err := maybeStartKubeLocalProxy(cf, withKubectlArgs(args), withKubeconfigEnvUpdate())
```